Update updatebeta

waltervl opened the pull request:
#4655 As discussed.
Tested with command line and with update through web interface on Linux 20.04 LTS.
Should be compatrible with RPi but please confirm.

waltervl committed:
Update updatebeta

rwaaren commented:
When executed it tar's the active database including the Write Ahead Log (wal) file and the SHaredMemory (shm)  file.
There is a (small) chance that these files, even when restored together, will result in a corrupt database.
See [how to corrupt a sqlite database](https://www.sqlite.org/howtocorrupt.html) -->  1.2. Backup or restore while a transaction is active

Safer methods would be to 

do the tar when no process access the database (= after domoticz stopped). 
or
before the tar do a sudo sqlite3 domoticz.db ".backup domoticz.backup" 
or
before the tar do a sudo sqlite3 domoticz.db "vacuum into 'domoticz.backup'"

waltervl committed:
Add stop start Domoticz

Stop Domoticz before creating backup to be sure to have a valid database backup

waltervl committed:
Update beta script

Stop Domoticz before creating backup to be sure to have a valid database backup

gizmocuz commented on updatebeta:
```diff
@@ -13,17 +13,39 @@ then
 if [ $? -ne 0 ]
 then
  echo "Problem in downloaded Domoticz archive. Stopping update!"
+ sudo service domoticz.sh start
```
This will leave you with a correct and partial Domoticz install.
Best to check if the wget command succeeded, if not, stop (and restart Domoticz)
You could check if the tar is correct, if not, stop (and restart Domoticz)
Extract , and restart Domoticz

But as mentioned, this script is also used internally and it will slow down the update proces quite a bit

Could you add a date/time timestamp in the filename as well ?

waltervl committed:
Added timestamp and extra check on wget

Time to create backup on my system is still into seconds. Might increase on slow hardware and large database. Total updatebeta time increases from 8 seconds to 10 seconds on my test system when started from commandline.

rwaaren reviewed (commented):
Tested on RPI Buster and NUC Debian Bullseye but 
script fails on both systems.


```
You are root.
Updating to latest beta version...
./updatebeta: line 55: syntax error near unexpected token `('
./updatebeta: line 55: `echo "Starting Domoticz... (please standby...)"'

```
Why changing to bash? It is not the same as sh and behaves different.
I assume ficho is a typo ?

waltervl committed:
typo

waltervl commented:
I changed it to bash because it is now the same way as the main installation script. Also I had to change it for the root access check (the same as the installation root access check),

rwaaren commented:
> I changed it to bash because it is now the same way as the main installation script. Also I had to change it for the root access check (the same as the installation root access check),

The main installation script is always executed from the command line . 

The updatebeta and updaterelease scripts can be executed from the command line and from the domoticz GUI. 

When executed from the command line I still get

```
You are root.
Updating to latest beta version...
./updatebeta: line 56: syntax error near unexpected token `('
./updatebeta: line 56: `echo "Starting Domoticz... (please standby...)"'

```

When updating from the GUI, I don't see any error message but I also see nothing happening in the domoticz log so I assume it is the same error.

waltervl committed:
Major copy action to remove all type errors

waltervl commented:
Sorry for this. First time working with github. 
I copied now the script as is that is on my system. This should work.

rwaaren commented:
> I copied now the script as is that is on my system. This should work.

Script now tested OK on NUC Debian bullseye  and RPI-4 buster. On slowest system it took 37 seconds in total ( 23 seconds extra compared to current updatebeta) when triggered from the GUI. 

Would it be considered useful to include version build number in backup name?
`build=$(./domoticz --version | grep Giz | awk '{print $5"_"$7}')
`

waltervl commented:
I do not know if version adds something, it will not hurt to add it.
What perhaps could be an issue is that there is only one backup. What if the user is updating, has issues, updates again the next day, still has issues. 
He cannot return to the original working environment. Is this a common scenario? And if yes, what is the max number of backups to keep?

rwaaren commented:
> I do not know if version adds something, it will not hurt to add it.
> What perhaps could be an issue is that there is only one backup. What if the user is updating, has issues, updates again the next day, still has issues.
> He cannot return to the original working environment. Is this a common scenario? And if yes, what is the max number of backups to keep?

My 2 cents; with this safeguard domoticz already takes a big part of the responsibility for failed updates up to a level that not many other software distributions do. I would leave it at 1 backup (at most)

waltervl commented:
So, let's merge it! Or do I have to do something urgent?

gizmocuz commented:
@waltervl , the modifications are good, but it introduces a huge delay in updating the application
At the present we have some other urgent matters to address, but I am thinking about adding a option in the settings that the user can enable to create backups for each update
So this PR will be used somehow

waltervl commented:
What is the problem with a huge delay for beta update as it it safe? I would not mind waiting 20 seconds extra.
Who is going to raise an issue? Better safe than sorry......
Why not just merge it (with an announcement on the forum) and see what the community thinks....

gizmocuz commented:
- I also need to do more testing... (different enviroments)
- test if we need this change "#!/usr/bin/env bash"
- and some prompts I would like to change

As mentioned this PR is not going to get wasted !

gizmocuz committed:
Updating some prompts

gizmocuz commented:
Question, at the beginning of the script you check if the user is root (or executed via sudo)
In the script you call sudo

I think we should either use sudo in the script, or check if it is executed with sudo rights and remove the sudo's in the script

What if Domoticz is not run as root ? sudo is used for stopping/starting the service
I know you did this for the backup archive creation, maybe use sudo there ?

gizmocuz commented:
Is it possible to create the backups in a/the 'backup' folder ?

waltervl commented:
> Question, at the beginning of the script you check if the user is root (or executed via sudo)
> In the script you call sudo
> 
> I think we should either use sudo in the script, or check if it is executed with sudo rights and remove the sudo's in the script
> 
> What if Domoticz is not run as root ? sudo is used for stopping/starting the service
> I know you did this for the backup archive creation, maybe use sudo there ?

Yes you are right, sudo should not be needed in the script itself as we check for that. 

As the standard install script installs Domoticz as root I kept that as default (the root check is taken from install).

waltervl commented:
> Is it possible to create the backups in a/the 'backup' folder ?

Yes, that should be no problem. But I do not know what happens if there is no backup folder and we create one for the this update and later the user decides to switch on automated database backup. New folders are created and perhaps this creation will stumble on the update backup folder. So I kept away from this.

gizmocuz committed:
Update updatebeta

removing bash dependency and small prompt changes

gizmocuz committed:
Added option to create no backup

gizmocuz committed:
Smoothing things...

gizmocuz commented:
Thank you for this PR! I will continue to tweak it a little bit more
